File: src/main/python/AoC2020_11.py
#! /usr/bin/env python3
#
# Advent of Code 2020 Day 11
#
import time
import os
from functools import lru_cache
from aoc import my_aocd


FLOOR = "."
EMPTY = "L"
OCCUPIED = "#"

the_grid: list[str]


@lru_cache(maxsize=8648)
def _find_adjacent(r: int, c: int,
                   num_rows: int, num_cols: int) -> tuple[tuple[int, int]]:
    return tuple((rr, cc)
                 for rr in range(r-1, r+2)
                 for cc in range(c-1, c+2)
                 if not (r == rr and c == cc)
                 and 0 <= rr < num_rows and 0 <= cc < num_cols)


@lru_cache(maxsize=8648)
def _find_visible(r: int, c: int,
                  num_rows: int, num_cols: int) -> list[tuple[int, int]]:
    global the_grid
    grid = the_grid
    vis = []
    for rg in [range(c-1, -1, -1), range(c+1, num_cols)]:
        for i in rg:
            if grid[r][i] != FLOOR:
                vis.append((r, i))
                break
    for rg in [range(r-1, -1, -1), range(r+1, num_rows)]:
        for i in rg:
            if grid[i][c] != FLOOR:
                vis.append((i, c))
                break
    cnt_right = num_cols-1-c
    cnt_up = r
    cnt_down = num_rows-1-r
    cnt_left = c
    i = 0
    while (i := i + 1) <= min(cnt_right, cnt_down):
        if grid[r+i][c+i] != FLOOR:
            vis.append((r+i, c+i))
            break
    i = 0
    while (i := i + 1) <= min(cnt_left, cnt_down):
        if grid[r+i][c-i] != FLOOR:
            vis.append((r+i, c-i))
            break
    i = 0
    while (i := i + 1) <= min(cnt_left, cnt_up):
        if grid[r-i][c-i] != FLOOR:
            vis.append((r-i, c-i))
            break
    i = 0
    while (i := i + 1) <= min(cnt_right, cnt_up):
        if grid[r-i][c+i] != FLOOR:
            vis.append((r-i, c+i))
            break
    return vis


def _run_cycle(grid: list[str], strategy, tolerance: int) -> list[str]:
    new_grid = list[str]()
    changed = False
    num_rows = len(grid)
    num_cols = len(grid[0])
    for r in range(num_rows):
        new_row = ""
        for c in range(num_cols):
            if grid[r][c] == EMPTY:
                to_check = strategy(r, c, num_rows, num_cols)
                if sum(1 for rr, cc in to_check
                       if grid[rr][cc] == OCCUPIED) == 0:
                    new_row += OCCUPIED
                    changed = True
                    continue
            elif grid[r][c] == OCCUPIED:
                to_check = strategy(r, c, num_rows, num_cols)
                if sum(1 for rr, cc in to_check
                       if grid[rr][cc] == OCCUPIED) >= tolerance:
                    new_row += EMPTY
                    changed = True
                    continue
            elif grid[r][c] != FLOOR:
                raise ValueError("invalid grid")
            new_row += grid[r][c]
        new_grid.append(new_row)
    return new_grid, changed


def _find_count_of_equilibrium(inputs: tuple[str],
                               strategy,
                               tolerance: int,
                               on_step=None) -> int:
    global the_grid
    the_grid = list(inputs)
    grid = the_grid
    while True:
        new_grid, changed = _run_cycle(grid, strategy, tolerance)
        if on_step is not None:
            on_step(new_grid)
        if not changed:
            return sum(row.count(OCCUPIED) for row in grid)
        grid = new_grid


def part_1(inputs: tuple[str]) -> int:
    return _find_count_of_equilibrium(inputs, _find_adjacent, tolerance=4)


def part_2(inputs: tuple[str]) -> int:
    return _find_count_of_equilibrium(inputs, _find_visible, tolerance=5)


def on_step(grid: list[str]) -> None:
    global step
    step += 1
    time.sleep(1)
    os.system('cls' if os.name in ('nt', 'dos') else 'clear')  # nosec
    print(f"Step {step}:")
    [print(line) for line in grid]


def visualize_part_1(inputs: tuple[str]) -> None:
    global step
    global the_grid
    step = 0
    os.system('cls' if os.name in ('nt', 'dos') else 'clear')  # nosec
    print("Initial:")
    [print(line) for line in the_grid]
    _find_count_of_equilibrium(inputs, _find_adjacent,
                               tolerance=4, on_step=on_step)


def visualize_part_2(inputs: tuple[str]) -> None:
    global step
    global the_grid
    step = 0
    os.system('cls' if os.name in ('nt', 'dos') else 'clear')  # nosec
    print("Initial:")
    [print(line) for line in the_grid]
    _find_count_of_equilibrium(inputs, _find_visible,
                               tolerance=5, on_step=on_step)


TEST = """\
L.LL.LL.LL
LLLLLLL.LL
L.L.L..L..
LLLL.LL.LL
L.LL.LL.LL
L.LLLLL.LL
..L.L.....
LLLLLLLLLL
L.LLLLLL.L
L.LLLLL.LL
""".splitlines()


def main() -> None:
    my_aocd.print_header(2020, 11)

    assert part_1(TEST) == 37
    assert part_2(TEST) == 26
    assert visualize_part_1(TEST) is None
    assert visualize_part_2(TEST) is None

    inputs = my_aocd.get_input(2020, 11, 94)
    result1 = part_1(inputs)
    print(f"Part 1: {result1}")
    result2 = part_2(inputs)
    print(f"Part 2: {result2}")


if __name__ == '__main__':
    main()

File: src/main/java/AoC2020_11.java
import static java.util.stream.Collectors.toSet;

import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.function.Function;
import java.util.stream.IntStream;
import java.util.stream.Stream;

import com.github.pareronia.aoc.Grid;
import com.github.pareronia.aoc.Grid.Cell;
import com.github.pareronia.aocd.Aocd;
import com.github.pareronia.aocd.Puzzle;

import lombok.RequiredArgsConstructor;

public class AoC2020_11 extends AoCBase {
	
    private static final char FLOOR = '.';
    private static final char EMPTY = 'L';
    private static final char OCCUPIED = '#';

	private final Grid grid;
	private final Map<Cell, Set<Cell>> adjacentCache;
	private final Map<Cell, Set<Cell>> visibleCache;
	
	private AoC2020_11(final List<String> input, final boolean debug) {
		super(debug);
		this.grid = Grid.from(input);
		this.adjacentCache = new HashMap<>();
		this.visibleCache = new HashMap<>();
	}
	
	public static AoC2020_11 create(final List<String> input) {
		return new AoC2020_11(input, false);
	}

	public static AoC2020_11 createDebug(final List<String> input) {
		return new AoC2020_11(input, true);
	}
	
	private Stream<Cell> adjacent(final Cell cell) {
	    return this.adjacentCache.computeIfAbsent(cell, this::findAdjacent).stream();
	}
	
	private Set<Cell> findAdjacent(final Cell cell) {
	    return IntStream.rangeClosed(cell.getRow() - 1, cell.getRow() + 1)
	            .filter(rr -> 0 <= rr && rr < this.grid.getHeight())
	            .boxed()
	            .flatMap(rr -> IntStream.rangeClosed(cell.getCol() - 1, cell.getCol() + 1)
	                            .filter(cc -> 0 <= cc && cc < this.grid.getWidth())
	                            .mapToObj(cc -> Cell.at(rr, cc)))
	            .filter(c -> !c.equals(cell))
	            .collect(toSet());
	}
	
	private Stream<Cell> visible(final Cell cell) {
	    return this.visibleCache.computeIfAbsent(cell, this::findVisible).stream();
	}
	
	private Set<Cell> findVisible(final Cell cell) {
	    final Set<Cell> cells = new HashSet<>();
	    for (final Stream<Cell> stream : Set.of(
	            this.grid.getCellsN(cell),
	            this.grid.getCellsNE(cell),
	            this.grid.getCellsE(cell),
	            this.grid.getCellsSE(cell),
	            this.grid.getCellsS(cell),
	            this.grid.getCellsSW(cell),
	            this.grid.getCellsW(cell),
	            this.grid.getCellsNW(cell)
	    )) {
	        stream.filter(c -> this.grid.getValueAt(c) != FLOOR).findFirst().ifPresent(cells::add);
	    }
	    return cells;
	}
	
	private long countOccupied(final Grid grid, final Cell cell, final Function<Cell, Stream<Cell>> strategy) {
	    return strategy.apply(cell).filter(c -> grid.getValueAt(c) == OCCUPIED).count();
	}
	
	private CycleResult runCycle(final Grid grid, final Function<Cell, Stream<Cell>> strategy, final int tolerance) {
	    final char[][] newGrid = new char[grid.getHeight()][grid.getWidth()];
	    boolean changed = false;
	    for (int rr = 0; rr < grid.getHeight(); rr++) {
	        final char[] newRow = new char[grid.getWidth()];
	        for (int cc = 0; cc < grid.getWidth(); cc++) {
	            final Cell cell = Cell.at(rr, cc);
                final char value = grid.getValueAt(cell);
                if (value == EMPTY && countOccupied(grid, cell, strategy) == 0) {
                    newRow[cc] = OCCUPIED;
                    changed = true;
	            } else if (value == OCCUPIED && countOccupied(grid, cell, strategy) >= tolerance) {
	                newRow[cc] = EMPTY;
	                changed = true;
	            } else  {
	                newRow[cc] = value;
	            }
	        }
	        newGrid[rr] = newRow;
	    }
	    return new CycleResult(new Grid(newGrid), changed);
	}
	
	private int findCountOfEquilibrium(final Function<Cell, Stream<Cell>> strategy, final int tolerance) {
	    Grid grid = this.grid;
	    while (true) {
	        final CycleResult result = runCycle(grid, strategy, tolerance);
	        if (!result.changed) {
	            return (int) result.grid.countAllEqualTo(OCCUPIED);
	        }
	        grid = result.grid;
	    }
	}
	
	@Override
	public Integer solvePart1() {
	    return findCountOfEquilibrium(this::adjacent, 4);
	}
	
	@Override
	public Integer solvePart2() {
	    return findCountOfEquilibrium(this::visible, 5);
	}

	public static void main(final String[] args) throws Exception {
		assert AoC2020_11.createDebug(TEST).solvePart1() == 37;
		assert AoC2020_11.createDebug(TEST).solvePart2() == 26;

        final Puzzle puzzle = Aocd.puzzle(2020, 11);
        puzzle.check(
            () -> lap("Part 1", () -> AoC2020_11.create(puzzle.getInputData()).solvePart1()),
            () -> lap("Part 2", () -> AoC2020_11.create(puzzle.getInputData()).solvePart2())
        );
	}
	
	private static final List<String> TEST = splitLines(
	        "L.LL.LL.LL\r\n" +
	        "LLLLLLL.LL\r\n" +
	        "L.L.L..L..\r\n" +
	        "LLLL.LL.LL\r\n" +
	        "L.LL.LL.LL\r\n" +
	        "L.LLLLL.LL\r\n" +
	        "..L.L.....\r\n" +
	        "LLLLLLLLLL\r\n" +
	        "L.LLLLLL.L\r\n" +
	        "L.LLLLL.LL"
	);
	
	@RequiredArgsConstructor
	private static final class CycleResult {
	    private final Grid grid;
	    private final boolean changed;
	}
}